File: Design pattern/AbstractFactory.py
from abc import ABCMeta, abstractmethod

class IProduce(metaclass=ABCMeta):
    def print_product(self):
        print(type(self).__name__)

class WoodChair(IProduce):
    pass

class PlasticChair(IProduce):
    pass

class WoodTable(IProduce):
    pass

class PlasticTable(IProduce):
    pass

class MaterialFactory(metaclass=ABCMeta):
    @abstractmethod
    def create_chair(self):
        pass

    @abstractmethod
    def create_table(self):
        pass

class WoodFactory(MaterialFactory):
    def create_chair(self):
        return WoodChair()

    def create_table(self):
        return WoodTable()

class PlasticFactory(MaterialFactory):
    def create_chair(self):
        return PlasticChair()

    def create_table(self):
        return PlasticTable()

class Factory:
    WOOD = 0
    PLASTIC = 1

    @staticmethod
    def get_material_factory(materialType) -> MaterialFactory:
        if materialType == Factory.WOOD:
            return WoodFactory()
        elif materialType == Factory.PLASTIC:
            return PlasticFactory()
        else:
            return None

if __name__ == "__main__":
    factory = Factory.get_material_factory(Factory.WOOD)
    product = factory.create_chair()
    product.print_product()
    product = factory.create_table()
    product.print_product()

    factory = Factory.get_material_factory(Factory.PLASTIC)
    product = factory.create_chair()
    product.print_product()
    product = factory.create_table()
    product.print_product()

File: Design pattern/AbstractFactory.cpp
#include <iostream>
using namespace std;

class IFuniture
{
public:
    virtual void print_funiture() = 0;
};

class WoodTable : public IFuniture
{
public:
    void print_funiture() override
    {
        cout << typeid(*this).name() << endl;
    }
};

class PlasticTable : public IFuniture
{
public:
    void print_funiture() override
    {
        cout << typeid(*this).name() << endl;
    }
};

class WoodChair : public IFuniture
{
public:
    void print_funiture() override
    {
        cout << typeid(*this).name() << endl;
    }
};

class PlasticChair : public IFuniture
{
public:
    void print_funiture() override
    {
        cout << typeid(*this).name() << endl;
    }
};

class IMaterialFactory
{
public:
    virtual IFuniture* create_chair() = 0;
    virtual IFuniture* create_table() = 0;
};

class WoodFactory : public IMaterialFactory
{
public:
    IFuniture* create_chair() override
    {
        return new WoodChair();
    }
    IFuniture* create_table() override
    {
        return new WoodTable();
    }
};

class PlasticFactory : public IMaterialFactory
{
public:
    IFuniture* create_chair() override
    {
        return new PlasticChair();
    }
    IFuniture* create_table() override
    {
        return new PlasticTable();
    }
};

class Factory
{
public:
    enum MaterialType
    {
        WOOD = 0,
        PLASTIC
    };
    static IMaterialFactory* get_factory(MaterialType type)
    {
        if (type == WOOD)
        {
            return new WoodFactory();
        }
        else if (type == PLASTIC)
        {
            return new PlasticFactory();
        }
        else
        {
            return nullptr;
        }
    }
};

int main()
{
    IMaterialFactory *factory;
    IFuniture *funiture;

    factory = Factory::get_factory(Factory::WOOD);
    funiture = factory->create_chair();
    funiture->print_funiture();
    delete funiture;
    funiture = factory->create_table();
    funiture->print_funiture();
    delete funiture;

    factory = Factory::get_factory(Factory::PLASTIC);
    funiture = factory->create_chair();
    funiture->print_funiture();
    delete funiture;
    funiture = factory->create_table();
    funiture->print_funiture();
    delete funiture;

    delete factory;
}